### PR TITLE
Add and fix export data types

### LINKED ISSUE #44: Numeric responses should be saved as numbers in XLS file
Currently all number responses in the Excel export are saved as text. Instead they should be saved as numbers (to allow the user to run calculations without having to manually change the data type each time).

image

### LINKED ISSUE #58: text and barcode response columns should be formatted as text in Excel
(otherwise leading zeros are stripped from e.g. phone numbers, ID numbers)

### LINKED ISSUE #59: Format dates so they are recognized by Excel
Dates are not recognized in Excel

yyyy-mm-dd isn't recognized by Excel (see current KC date exports which works in all locales)

### LINKED ISSUE #119: decimal question fails autoreport

### LINKED ISSUE #243: Handle pyxform v1.5.0 features
- Step 1: ensure exports succeed with new pxform features
- Step 2: integrate features properly into exports

### LINKED ISSUE #253: `_validation_status` value not carried over to rows in repeat groups in XLS export
## Description

1. Create a form with repeat group
2. Submit data
3. Change validation status of submission
3. Export from KPI
4. Repeat group sheet will not contain the values of the submissions' `_validation_status` in their `_submission__validation_status` column

### PR DESCRIPTION
**Pull request by joshuaberetta**
Adapted table from https://github.com/kobotoolbox/formpack/issues/58#issuecomment-801433613
question type | store data as | display as
---|---|---
integer | general | number
decimal | general | number
range | general | number
text | general | text
select_one | general | number or text
select_one_from_file | general | number or text
select_multiple | general | number or text
select_multiple_from_file | general | text
select_multiple (split columns) | general | number
rank | general | text
geopoint | general | text
geopoint (split columns) | general | number
geotrace | general | text
geoshape | general | text
date | date | date
time* | general | text
dateTime | date | date
today | date | date
start | date | date
end | date | date
barcode | general | text
calculate | general | text

\* For `time` to be exported as a `Time` Excel type, we need to restructure how we handle the exports to explicitly determine the type for each cell. Currently XlsxWriter determines the Excel type based on the Python data type when using `write_row()` and unfortunately treats `datetime.datetime` the same as `datetime.time` when a `default_date_format` is passed to the `xlsxwriter.Workbook`.

## Description

Facilitate Excel data types for numbers and dates in exports.

## Related issues:

closes #44 
closes #58 
closes #59 
possibly close #119, #137
closes #253 
kpi support in kobotoolbox/kpi#3330

### CODE REVIEW

File: src/formpack/reporting/export.py [hunk 1]
```diff
@@ -731,17 +729,13 @@ def to_xlsx(self, filename, submissions):
         sheet_row_positions = defaultdict(lambda: 0)
 
         def _append_row_to_sheet(sheet_, data):
-            # Ensure all list objects are coerced to strings otherwise
-            # xlswriter will fail to export
-            data_ = [str(d) if isinstance(d, list) else d for d in data]
-
```
Comment (joshuaberetta) [inline src/formpack/reporting/export.py:737]: Now handled in the `TagsCopyField` and `NotesCopyField` classes.

File: src/formpack/reporting/export.py [hunk 2]
```diff
@@ -390,15 +391,6 @@ def format_one_submission(self, submission, current_section):
...
 
-            # If the submission has been tagged, join those tags together into
-            # a comma-separated string
-            for tags_col in ('_tags', '_submission__tags'):
-                if tags_col in row:
-                    tags = row[tags_col]
-                    row[tags_col] = (
-                        ', '.join(tags) if isinstance(tags, list) else tags
-                    )
-
```
Comment (joshuaberetta) [inline src/formpack/reporting/export.py:401]: Now handled in the `TagsCopyField` class.

File: src/formpack/reporting/export.py [hunk 3]
```diff
@@ -80,6 +83,13 @@ def __init__(self, formpack, form_versions, lang=UNSPECIFIED_TRANSLATION,
                         dumb_field = CopyField(copy_field, section=first_section)
                     first_section.fields[dumb_field.name] = dumb_field
 
+        # Some copy fields are classes, some strings -- collect their field
+        # names for later use
+        self.copy_field_names = [
+            getattr(copy_field, 'FIELD_NAME', copy_field)
+            for copy_field in self.copy_fields
+        ]
+
```
Comment (joshuaberetta) [inline src/formpack/reporting/export.py:92]: addresses formpack#253

File: src/formpack/reporting/export.py [hunk 4]
```diff
@@ -721,7 +726,14 @@ def to_table(self, submissions):
         return table
 
     def to_xlsx(self, filename, submissions):
-        workbook = xlsxwriter.Workbook(filename, {'constant_memory': True})
+        workbook = xlsxwriter.Workbook(
+            filename,
+            {
+                'constant_memory': True,
+                'default_date_format': 'yyyy-mm-dd',
+                'remove_timezone': True,
```
Comment (joshuaberetta) [inline src/formpack/reporting/export.py:734]: Required when exporting dates as `datetime` objects

File: src/formpack/schema/fields.py [hunk 5]
```diff
@@ -179,15 +181,16 @@ def from_json_definition(cls, definition, hierarchy=None,
...
-            'end': TextField,
+            'datetime': DateTimeField,
+            'start': DateTimeField,
+            'end': DateTimeField,
 
             # general
             'text': TextField,
             'barcode': TextField,
             'acknowledge': TextField,
+            'calculate': TextField,
```
Comment (joshuaberetta) [inline src/formpack/schema/fields.py:193]: Keep as `TextField` based on table here: https://github.com/kobotoolbox/formpack/issues/58#issuecomment-801433613

File: src/formpack/schema/fields.py [hunk 6]
```diff
@@ -264,6 +267,20 @@ def get_disaggregated_stats(self, metrics, top_splitters,
     def parse_values(self, raw_values):
         yield raw_values
 
+    @staticmethod
+    def try_get_number(val):
+        try:
+            val = int(val)
+        except ValueError:
+            pass
+        if isinstance(val, int):
```
Comment (noliveleger) [inline src/formpack/schema/fields.py:276]: Would `else` be enough (instead of `if isinstance(val, int):`)?

File: src/formpack/schema/fields.py [hunk 7]
```diff
@@ -452,6 +469,41 @@ def get_disaggregated_stats(self, metrics, top_splitters,
...
+
+        if not xls_types:
+            return {self.name: val}
+
+        _date = val
+        try:
+            _date = parser.parse(val)
+        except ValueError:
+            pass
+        if isinstance(_date, datetime):
```
Comment (noliveleger) [inline src/formpack/schema/fields.py:484]: Same comment about `else`? 

if `parser.parse` works as expected, it returns a `datetime` object. 

```
        try:
            _date = parser.parse(val)
        except ValueError:
            pass
        else:
             # This assertion should be always True
             # assert type(_date) == datetime
             _date = _date.date()
                   
```        

Did I miss something?

File: src/formpack/schema/fields.py [hunk 8]
```diff
@@ -560,6 +621,96 @@ def get_labels(self, *args, **kwargs):
         return [self.name]
 
 
+class IdCopyField(CopyField):
+
+    FIELD_NAME = '_id'
+
+    def __init__(self, section=None, *args, **kwargs):
+        super(IdCopyField, self).__init__(
```
Comment (noliveleger) [inline src/formpack/schema/fields.py:629]: We can drop Python2 support. ( I though we did :-) )

File: src/formpack/schema/fields.py [hunk 9]
```diff
@@ -560,6 +621,96 @@ def get_labels(self, *args, **kwargs):
...
+
+        return {self.name: int(val)}
+
+
+class SubmissionTimeCopyField(CopyField):
+
+    FIELD_NAME = '_submission_time'
+
+    def __init__(self, section=None, *args, **kwargs):
+        super(SubmissionTimeCopyField, self).__init__(
```
Comment (noliveleger) [inline src/formpack/schema/fields.py:651]: Drop Python2 support.

File: src/formpack/schema/fields.py [hunk 10]
```diff
@@ -560,6 +621,96 @@ def get_labels(self, *args, **kwargs):
...
+
+    def format(self, val, *args, **kwargs):
+        if val and isinstance(val, list):
+            val_ = ', '.join(val)
+        else:
+            val_ = ''
+
+        return {self.name: val_}
+
+
```
Comment (noliveleger) [inline src/formpack/schema/fields.py:713]: 👏

**Review by noliveleger — CHANGES_REQUESTED** [review]

File: src/formpack/schema/fields.py (hunk 8 above)
Comment (noliveleger) [reply src/formpack/schema/fields.py:629]: It seems it was only in my dreams. 
https://github.com/kobotoolbox/formpack/issues/239

File: src/formpack/schema/fields.py (hunk 6 above)
Comment (joshuaberetta) [reply src/formpack/schema/fields.py:276]: Ah beautiful ❤️

```python3
In [1]: def get_num(num):
   ...:     try:
   ...:         num = int(num)
   ...:     except ValueError:
   ...:         pass
   ...:     else:
   ...:         return type(num)
   ...:     try:
   ...:         num = float(num)
   ...:     except ValueError:
   ...:         pass
   ...:     return type(num)
   ...:

In [2]: get_num('42')
Out[2]: int

In [3]: get_num('42.42')
Out[3]: float

In [4]: get_num('42.42a')
Out[4]: str
```

File: src/formpack/schema/fields.py (hunk 7 above)
Comment (joshuaberetta) [reply src/formpack/schema/fields.py:484]: You missed nothing 💁

**Pull request merged**